import osfrom Bio import SeqIOfrom Bio.Blast.Applications import NcbiblastnCommandline as blastnimport pandas as pdfrom io import StringIOfrom math import isnanls = os.listdir()db = "/home/avian-flu/Desktop/Ag_Shift/kp/blast_db/region"num_threads = 32outfmt = "6 sseqid qstart qend"def blast(file):    test=blastn(db=db, query=file, num_threads=num_threads, outfmt=outfmt)    test_df = pd.read_csv(StringIO(test()[0]),sep="\t",header=None,names=['gene','start','stop'])    df = test_df.sort_values('start').reset_index().drop('index',axis=1)    seq_min = df['start'].min()    seq_max = df['stop'].max()    return {file:[seq_min,seq_max]}blast_dict = dict()for i in ls:    blast_dict.update(blast(i))for i in blast_dict.keys():    if isnan(blast_dict[i][0]) != True:        SeqIO.write(SeqIO.read(i,"fasta")[blast_dict[i][0]-1:blast_dict[i][1]-1],"../part/part_{}".format(i),"fasta")from Bio import Entrezdef ider(key):    term = key.replace(".fna","")    ids = Entrez.read(handle = Entrez.esearch(db="nucleotide", term=term, rettype=id))['IdList']    return ids# gb downloadfor i in blast_dict.keys():    term = i.replace(".fna","")    SeqIO.write(SeqIO.read(Entrez.efetch(db="nucleotide", id=term, rettype="gb", seq_start=blast_dict[i][0], seq_stop=blast_dict[i][1]),"gb"), "gb/{}.gb".format(term), "gb")